CRAN snprintf warnings

Version: 0.9
Check: whether package can be installed
Result: WARN 
  Found the following significant warnings:
    modiSBML.c:492:37: warning: ‘%s’ directive output may be truncated writing up to 5118 bytes into a region of size between 1 and 5119 [-Wformat-truncation=]
    modiSBML.c:578:37: warning: ‘%s’ directive output may be truncated writing up to 5118 bytes into a region of size between 1 and 5119 [-Wformat-truncation=]
  See ‘/home/hornik/tmp/R.check/r-devel-gcc/Work/PKGS/MCSimMod.Rcheck/00install.out’ for details.
  * used C compiler: ‘gcc-14 (Debian 14.2.0-19) 14.2.0’
Flavor: [r-devel-linux-x86_64-debian-gcc](https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-debian-gcc/MCSimMod-00check.html)

Version: 0.9
Check: whether package can be installed
Result: WARN 
  Found the following significant warnings:
    modiSBML.c:492:37: warning: ‘%s’ directive output may be truncated writing up to 5118 bytes into a region of size between 1 and 5119 [-Wformat-truncation=]
    modiSBML.c:578:37: warning: ‘%s’ directive output may be truncated writing up to 5118 bytes into a region of size between 1 and 5119 [-Wformat-truncation=]
  See ‘/data/gannet/ripley/R/packages/tests-devel/MCSimMod.Rcheck/00install.out’ for details.
  * used C compiler: ‘gcc (GCC) 14.2.1 20240912 (Red Hat 14.2.1-3)’
Flavor: [r-devel-linux-x86_64-fedora-gcc](https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-fedora-gcc/MCSimMod-00check.html)

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -19,7 +19,7 @@ jobs:
       fail-fast: false
       matrix:
         config:
-          - {os: ubuntu-22.04, r: 'release'}
+          - {os: ubuntu-24.04, r: 'release'} #ensures we are running with a recent version of gcc similar to CRAN
           - {os: windows-2022, r: 'release'}
     env:
       R_KEEP_PKG_SOURCE: yes

--- a/src/modiSBML.c
+++ b/src/modiSBML.c
@@ -481,7 +481,7 @@ __attribute__((warn_unused_result)) int Transcribe1AlgEqn(PFILE pfile, PVMMAPSTR
     PROPAGATE_EXIT(NextLex(&ibDummy, szLex, &iType)); /* ...all errors reported */
 
     if ((iType == LX_IDENTIFIER) && !(IsMathFunc(szLex)) && (szLex[0] == '_')) {
-      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1;
+      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1; //
       if (len > MAX_EQN) {
         Rprintf("\n***Error: max string length MAX_EQN exceeded in "
                 "Transcribe1AlgEqn: %s%s%s\n",
@@ -489,8 +489,10 @@ __attribute__((warn_unused_result)) int Transcribe1AlgEqn(PFILE pfile, PVMMAPSTR
         Rprintf("Exiting...\n\n");
         return EXIT_ERROR;
       }
-      snprintf(szTmpEqSwap, len, "%s%s%s", szTmpEq, pV->szName, szLex); /* extend */
-      strncpy(szTmpEq, szTmpEqSwap, len);
+      // because of the check above we ensure we will be under MAX_EQN 
+      // so the following operations will always emit non negative values
+      strncat(szTmpEq, pV->szName, MAX_EQN - strlen(szTmpEq) - 1);
+      strncat(szTmpEq, szLex, MAX_EQN - strlen(szTmpEq) - 1);
     } else {
       int len = strlen(szTmpEq) + 1 + strlen(szLex) + 1;
       if (len > MAX_EQN) {
@@ -567,7 +569,7 @@ __attribute__((warn_unused_result)) int Transcribe1DiffEqn(PFILE pfile, PVMMAPST
     PROPAGATE_EXIT(NextLex(&ibDummy, szLex, &iType)); /* ...all errors reported */
 
     if ((iType == LX_IDENTIFIER) && !(IsMathFunc(szLex)) && (szLex[0] == '_')) {
-      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1;
+      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1; 
       if (len > MAX_EQN) {
         Rprintf("\n***Error: max string length MAX_EQN exceeded in "
                 "Transcribe1DiffEqn: %s%s%s\n",
@@ -575,8 +577,11 @@ __attribute__((warn_unused_result)) int Transcribe1DiffEqn(PFILE pfile, PVMMAPST
         Rprintf("Exiting...\n\n");
         return EXIT_ERROR;
       }
-      snprintf(szTmpEqSwap, len, "%s%s%s", szTmpEq, pV->szName, szLex); /* extend */
-      strncpy(szTmpEq, szTmpEqSwap, len);
+      // because of the check above we ensure we will be under MAX_EQN 
+      // so the following operations will always emit non negative values
+      strncat(szTmpEq, pV->szName, MAX_EQN - strlen(szTmpEq) - 1);
+      strncat(szTmpEq, szLex, MAX_EQN - strlen(szTmpEq) - 1);
+
     } else {
       int len = strlen(szTmpEq) + 1 + strlen(szLex) + 1;
       if (len > MAX_EQN) {

--- a/src/modiSBML.c
+++ b/src/modiSBML.c
@@ -481,7 +481,7 @@ __attribute__((warn_unused_result)) int Transcribe1AlgEqn(PFILE pfile, PVMMAPSTR
     PROPAGATE_EXIT(NextLex(&ibDummy, szLex, &iType)); /* ...all errors reported */
 
     if ((iType == LX_IDENTIFIER) && !(IsMathFunc(szLex)) && (szLex[0] == '_')) {
-      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1; //
+      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1;
       if (len > MAX_EQN) {
         Rprintf("\n***Error: max string length MAX_EQN exceeded in "
                 "Transcribe1AlgEqn: %s%s%s\n",
@@ -569,7 +569,7 @@ __attribute__((warn_unused_result)) int Transcribe1DiffEqn(PFILE pfile, PVMMAPST
     PROPAGATE_EXIT(NextLex(&ibDummy, szLex, &iType)); /* ...all errors reported */
 
     if ((iType == LX_IDENTIFIER) && !(IsMathFunc(szLex)) && (szLex[0] == '_')) {
-      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1; 
+      int len = strlen(szTmpEq) + strlen(pV->szName) + strlen(szLex) + 1;
       if (len > MAX_EQN) {
         Rprintf("\n***Error: max string length MAX_EQN exceeded in "
                 "Transcribe1DiffEqn: %s%s%s\n",

